### PR TITLE
tests: Refactor SR-IOV tests

### PR DESCRIPTION
**Pull request by EdDev**
**What this PR does / why we need it**:

Refactor and simplify the e2e SR-IOV tests.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: tests/network/sriov.go [hunk 1]
```diff
@@ -70,11 +70,7 @@ var _ = Describe("[Serial]SRIOV", func() {
 
 	var virtClient kubecli.KubevirtClient
 
-	sriovResourceName := os.Getenv("SRIOV_RESOURCE_NAME")
-
-	if sriovResourceName == "" {
-		sriovResourceName = "kubevirt.io/sriov_net"
-	}
+	sriovResourceName := readSRIOVResourceName()
```
Comment (orelmisan) [inline tests/network/sriov.go:73]: Can you please move this change to a commit of its own?
I find it different from this commit's subject.

File: tests/network/sriov.go [hunk 2]
```diff
@@ -182,8 +159,9 @@ var _ = Describe("[Serial]SRIOV", func() {
 			defer tests.UpdateKubeVirtConfigValueAndWait(testsuite.KubeVirtDefaultConfig)
 
 			vmi := newSRIOVVmi([]string{sriovnet1}, defaultCloudInitNetworkData())
-			vmi = startVmi(vmi)
-			vmi = waitVmi(vmi)
+			var err error
+			vmi, err = createAndWaitVMI(vmi)
```
Comment (orelmisan) [inline tests/network/sriov.go:103]: nit: Is it possible to use a short variable declaration?
There are several similar cases bellow.

File: tests/network/sriov.go [hunk 3]
```diff
@@ -697,3 +684,30 @@ func checkInterfacesInGuest(vmi *v1.VirtualMachineInstance, interfaces []string)
 		Expect(checkInterface(vmi, iface)).To(Succeed())
 	}
 }
+
+// createAndWaitVMI creates the received VMI and waits for the guest to load the guest-agent
+func createAndWaitVMI(vmi *v1.VirtualMachineInstance) (*v1.VirtualMachineInstance, error) {
```
Comment (orelmisan) [inline tests/network/sriov.go:669]:
```suggestion
func createVMIAndWait(vmi *v1.VirtualMachineInstance) (*v1.VirtualMachineInstance, error) {
```

File: tests/network/sriov.go [hunk 4]
```diff
@@ -697,16 +697,21 @@ func createAndWaitVMI(vmi *v1.VirtualMachineInstance) (*v1.VirtualMachineInstanc
 		return nil, err
 	}
 
-	return waitVMI(virtClient, vmi)
+	return waitVMI(vmi)
 }
 
-func waitVMI(virtClient kubecli.KubevirtClient, vmi *v1.VirtualMachineInstance) (*v1.VirtualMachineInstance, error) {
+func waitVMI(vmi *v1.VirtualMachineInstance) (*v1.VirtualMachineInstance, error) {
```
Comment (orelmisan) [inline tests/network/sriov.go:683]: Can you please add the reasoning for this change in the commit message?

**Review by orelmisan — COMMENTED** [review]
Thank you for the PR @EdDev.
Please see the inline comments.

File: tests/network/sriov.go (hunk 1 above)
Comment (EdDev) [reply tests/network/sriov.go:73]: Done

File: tests/network/sriov.go (hunk 2 above)
Comment (EdDev) [reply tests/network/sriov.go:103]: Could you please explain what you mean? Perhaps by an example?

File: tests/network/sriov.go (hunk 3 above)
Comment (EdDev) [reply tests/network/sriov.go:669]: Done

File: tests/network/sriov.go (hunk 4 above)
Comment (EdDev) [reply tests/network/sriov.go:683]: Do I really need to explain why it is simpler?
As I need to add other changes, I do not mind much, but this one seems trivial enough to me.
Done.

File: tests/network/sriov.go (hunk 2 above)
Comment (orelmisan) [reply tests/network/sriov.go]: Is this code snippet possible in this location?
```go
vmi, err := createVMIAndWait(vmi)
```

File: tests/network/sriov.go (hunk 4 above)
Comment (orelmisan) [reply tests/network/sriov.go:683]: I did not know that getting the client more then once has little cost, this is the reason I've asked.
Thanks for the explanation.

File: tests/network/sriov.go (hunk 2 above)
Comment (orelmisan) [reply tests/network/sriov.go]: Instead of the declaration of the `err` variable:
```go
var err error
```

Comment (EdDev) [reply tests/network/sriov.go]: No, because `vmi` is already declared.

Comment (EdDev) [reply tests/network/sriov.go]: Well, I think it will shadow it.
I do not know if that is problematic or not. Most usages do not attempt to do this. If you have some ref about this pattern, we can try and consider it.

Comment (orelmisan) [reply tests/network/sriov.go]: It is fine that `vmi` already exists, as long as `err` is a new variable that is being introduced.
Please see https://go.dev/ref/spec#Short_variable_declarations .

Comment (orelmisan) [reply tests/network/sriov.go]: > Unlike regular variable declarations, a short variable declaration may redeclare variables provided they were originally declared earlier in the same block (or the parameter lists if the block is the function body) with the same type, and at least one of the non-[blank](https://go.dev/ref/spec#Blank_identifier) variables is new.

Comment (EdDev) [reply tests/network/sriov.go]: Yes, seems like it is ok.
Even if it was a new variable shadowing the original, in this case it would have been fine as the variable is not bounded to another function/block.

I do not really like this odd definition, it causes to many question marks to appear.
But I cannot fight the spec, nor the fact it does work (passes build & linter).

Done

File: tests/network/sriov.go [hunk 5]
```diff
@@ -92,514 +88,453 @@ var _ = Describe("[Serial]SRIOV", func() {
...
 
-				checkDefaultInterfaceInPod(vmi)
+			BeforeEach(func() {
+				// The SR-IOV VF MAC should be preserved on migration, therefore explicitly specify it.
+				vmi = newSRIOVVmi([]string{sriovnet1}, defaultCloudInitNetworkData())
+				vmi.Spec.Domain.Devices.Interfaces[1].MacAddress = mac
 
-				By("checking virtual machine instance has two interfaces")
-				checkInterfacesInGuest(vmi, []string{"eth0", "eth1"})
+				vmi, err := createVMIAndWait(vmi)
```
Comment (EdDev) [inline tests/network/sriov.go:353]: This has now shadowed the vmi variable, causing the test not to get this last update.

Comment (EdDev) [reply tests/network/sriov.go]: Fixed

**Comment by EdDev** [comment]
[change](https://github.com/kubevirt/kubevirt/compare/7c5f4a8d2786928b24654381898cfe817eba80f5..a399fe34f8e30c30be2173942b0eb7e6b432b3e5): Answered review comments

**Review by orelmisan — APPROVED** [review]
@EdDev Thank you for the changes.

File: tests/network/sriov.go [hunk 6]
```diff
@@ -575,6 +571,15 @@ var _ = Describe("[Serial]SRIOV", func() {
 	})
 })
 
+func readSRIOVResourceName() string {
+	sriovResourceName := os.Getenv("SRIOV_RESOURCE_NAME")
+	if sriovResourceName == "" {
+		const defaultTestSRIOVResourceName = "kubevirt.io/sriov_net"
```
Comment (RamLavi) [inline tests/network/sriov.go:533]: can you explain why do you define the const in this scope, and not outside?

**Review by RamLavi — COMMENTED** [review]
The diff in commit `tests, sriov: Drop one node (context) layer` is not working well - did you only remove the layer?

File: tests/network/sriov.go (hunk 6 above)
Comment (EdDev) [reply tests/network/sriov.go:533]: I follow the recommendation to define vars/consts as close as possible to the usage.
As a side effect, you now know that no one else is using this const.

**Comment by EdDev** [comment]
> The diff in commit `tests, sriov: Drop one node (context) layer` is not working well - did you only remove the layer?

You can configure the diff to ignore whitespaces and then it should look fine.

File: tests/network/sriov.go [hunk 7]
```diff
@@ -461,7 +431,15 @@ var _ = Describe("[Serial]SRIOV", func() {
 			Expect(err).ToNot(HaveOccurred())
 
 			//create two vms on the same sriov network
-			vmi1, vmi2 := createSriovVMs(sriovnetLinkEnabled, sriovnetLinkEnabled, cidrA, cidrB)
+			vmi1, err := createSRIOVVmiOnNode(sriovNode, sriovnetLinkEnabled, cidrA)
+			Expect(err).ToNot(HaveOccurred())
+			vmi2, err := createSRIOVVmiOnNode(sriovNode, sriovnetLinkEnabled, cidrB)
+			Expect(err).ToNot(HaveOccurred())
+
+			vmi1, err = waitVMI(vmi1)
```
Comment (RamLavi) [inline tests/network/sriov.go:439]: why not do `createSRIOVVmiOnNodeAndWait` like done in previous commit?

File: tests/network/sriov.go [hunk 8]
```diff
@@ -695,3 +670,28 @@ func findIfaceByMAC(virtClient kubecli.KubevirtClient, vmi *v1.VirtualMachineIns
...
+	}
+	if cloudInitNetworkData != "" {
+		cloudinitOption := libvmi.WithCloudInitNoCloudNetworkData(cloudInitNetworkData)
+		options = append(options, cloudinitOption)
+	}
+
+	for _, name := range networks {
+		options = append(options,
+			libvmi.WithInterface(libvmi.InterfaceDeviceWithSRIOVBinding(name)),
+			libvmi.WithNetwork(libvmi.MultusNetwork(name, name)),
```
Comment (RamLavi) [inline tests/network/sriov.go:648]: a bit of an off topic, so I'll leave it as suggestion only: using same name for NAD and network name does not allow us to run tests with 2 NICS connected to same NAD, is that something you want to address in this PR as well?

File: tests/network/sriov.go (hunk 6 above)
Comment (RamLavi) [reply tests/network/sriov.go:533]: then why not just use:
```
sriovResourceName = "kubevirt.io/sriov_net"
```
Why do we have to go through const? seems odd

Comment (RamLavi) [reply tests/network/sriov.go:533]: Let me rephrase
we use const so that no one could change the var in the future..
but here you create a const and immediately assign the value to a (non-const) variable... 
if the const has no further use and the variable using it is [mutable](https://go.dev/play/p/GqPWnCbDH5k), then why use the const in the first place?

Comment (EdDev) [reply tests/network/sriov.go:533]: > Let me rephrase we use const so that no one could change the const in the future..

A const is processed at compilation and as such has several benefits, one of which is that it is immutable.
But I think one of its other usages is to name scalars and strings, giving them meaning beyond their value.

> but here you create a const and immediately assign the value to a (non-const) variable... if the const has no further use and the variable using it is [mutable](https://go.dev/play/p/GqPWnCbDH5k), then why use the const in the first place?

Because we are interested in explaining what is this `"kubevirt.io/sriov_net"` representing.
We can set in the variable mentioned either this default or a value from an env variable.

File: tests/network/sriov.go (hunk 7 above)
Comment (EdDev) [reply tests/network/sriov.go:439]: The original `createSriovVMs` has created two VM/s and only then waited from them.
This is doing the same, starting the VM/s in parallel and waiting for both afterwards.

File: tests/network/sriov.go (hunk 8 above)
Comment (EdDev) [reply tests/network/sriov.go:648]: This scenario has not existed so far and it was not there even in the previous helper.
I think we should adjust it accordingly when the need lands.
At that time, we may need a further refactoring done here (e.g. split between single secondary network and multi secondary networks).

File: tests/network/sriov.go [hunk 9]
```diff
@@ -689,16 +689,21 @@ func createVMIAndWait(vmi *v1.VirtualMachineInstance) (*v1.VirtualMachineInstanc
...
+func waitVMI(vmi *v1.VirtualMachineInstance) (*v1.VirtualMachineInstance, error) {
 	// Running multi sriov jobs with Kind, DinD is resource extensive, causing DeadlineExceeded transient warning
 	// Kubevirt re-enqueue the request once it happens, so its safe to ignore this warning.
 	// see https://github.com/kubevirt/kubevirt/issues/5027
 	warningsIgnoreList := []string{"unknown error encountered sending command SyncVMI: rpc error: code = DeadlineExceeded desc = context deadline exceeded"}
 	tests.WaitUntilVMIReadyIgnoreSelectedWarnings(vmi, console.LoginToFedora, warningsIgnoreList)
 
+	virtClient, err := kubecli.GetKubevirtClient()
+	if err != nil {
+		panic(err)
```
Comment (ormergi) [inline tests/network/sriov.go:684]: Why panic and not return an err? 
Returning an error may give a chance to other tests to pass instead of stopping the whole test suite.

File: tests/network/sriov.go [hunk 10]
```diff
@@ -719,3 +721,40 @@ func checkDefaultInterfaceInPod(vmi *v1.VirtualMachineInstance) error {
...
+	vmi := newSRIOVVmi([]string{networkName}, cloudInitNetworkDataWithStaticIPsByMac(networkName, mac.String(), cidr))
+	const secondaryInterfaceIndex = 1
+	vmi.Spec.Domain.Devices.Interfaces[secondaryInterfaceIndex].MacAddress = mac.String()
+
+	vmi = tests.CreateVmiOnNode(vmi, nodeName)
+
+	return vmi, nil
+}
+
+func sriovNodeName(sriovResourceName string) (string, error) {
```
Comment (ormergi) [inline tests/network/sriov.go:749]: I would change this function to return a list of nodes and pick one of them at the caller,
and rename it to `listNodeByResource`.
It may help the reader understand where VMs are created.

**Review by ormergi — COMMENTED** [review]
Great work!

File: tests/network/sriov.go (hunk 9 above)
Comment (EdDev) [reply tests/network/sriov.go:684]: Ginkgo `Fail` also panics, so tests will continue running even with such a direct panic.

You should check the original code and see it also panics here.
The reasoning is that errors in tests better be differentiated from failures in expectations (assertions).

File: tests/network/sriov.go (hunk 10 above)
Comment (EdDev) [reply tests/network/sriov.go:749]: The function abstracts how the node is selected. The reader at the caller side should either be satisfied with understanding that a suitable node is returned or, if he insist, can go into the function implementation and examine how it is done. Exposing the list reduces the abstraction and requires the caller to do some logic to chose one.

I think the existing proposal is better.

File: tests/network/sriov.go (hunk 8 above)
Comment (RamLavi) [reply tests/network/sriov.go:648]: yeah totally, just wanted to mention that it's something we may want to tackle in the future..

File: tests/network/sriov.go (hunk 7 above)
Comment (RamLavi) [reply tests/network/sriov.go:439]: just saying it would reduce caller code :) ..

File: tests/network/sriov.go (hunk 6 above)
Comment (RamLavi) [reply tests/network/sriov.go:533]: if they have meaning beyond their value, then shouldn't it live in the broader scope of the test?

Looking at the test code, I don't think we ever do a `os.Setenv("SRIOV_RESOURCE_NAME", value)` so aren't we always getting into this if condition?

Anyway I don't think defining a const in this example and scope is contributing anything, I would add the const in the broader scope and use it here (and perhaps later as well).

File: tests/network/sriov.go (hunk 7 above)
Comment (EdDev) [reply tests/network/sriov.go:439]: Reducing caller code vs reducing test time. Usually the later wins, don't you agree?

File: tests/network/sriov.go (hunk 6 above)
Comment (EdDev) [reply tests/network/sriov.go:533]: > if they have meaning beyond their value, then shouldn't it live in the broader scope of the test?

If no one else uses it directly, why would we be interested in increasing its scope?
It expresses the reality, nothing more.
I would expect no one to need it at the moment, just to consume this helper.
If that changes in the future, then the scope needs to change accordingly.

> Looking at the test code, I don't think we ever do a os.Setenv("SRIOV_RESOURCE_NAME", value) so aren't we always getting into this if condition?

I think the idea is for the one who is executing the tests, to control this per the setup they have.

> Anyway I don't think defining a const in this example and scope is contributing anything, I would add the const in the broader scope and use it here (and perhaps later as well).

I answered on why the scope IMO is fine.
And I think it helps readers (including me) to express what this string is.

**Review by ormergi — APPROVED** [review]

**Comment by EdDev** [comment]
The[ job failure](https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/8299/pull-kubevirt-e2e-k8s-1.23-sig-network-nonroot/1559759557104766976) is not related.
 
/test pull-kubevirt-e2e-k8s-1.23-sig-network-nonroot
